networkx/algorithms/community/local.py

```diff
@@ -0,0 +1,223 @@
+
+
+"""
+
+__all__ = ["greedy_source_expansion"]
+
+
+def _clauset_greedy_source_expansion(G, *, source, cutoff=None):
+    if cutoff is None:
+        cutoff = float("inf")
+    C = {source}
+    B = {source}
+    U = G[source].keys() - C
+    T = {frozenset([node, neighbor]) for node in B for neighbor in G.neighbors(node)}
+    I = {edge for edge in T if all(node in C for node in edge)}
+
+    R_value = 0
+    while len(C) < cutoff:
+        if not U:
+            break
+
+        max_R = 0
+        best_node = None
+        best_node_B = best_node_T = best_node_I = set()
+
+        for v in U:
+            R_tmp, B_tmp, T_tmp, I_tmp = _calculate_local_modularity_for_candidate(
+                G, v, C, B, T, I
+            )
+            if R_tmp > max_R:
+                max_R = R_tmp
+                best_node = v
+                best_node_B = B_tmp
+                best_node_T = T_tmp
+                best_node_I = I_tmp
+
+        C = C | {best_node}
+        U.update(G[best_node].keys() - C)
+        U.remove(best_node)
+        B = best_node_B
+        T = best_node_T
+        I = best_node_I
+        if max_R < R_value:
+            break
+        R_value = max_R
+
+    return C
+
+
+def _calculate_local_modularity_for_candidate(G, v, C, B, T, I):
+    """
+    Compute the local modularity R and updated variables when adding node v to the community.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        The input graph.
+    v : node
+        The candidate node to add to the community.
+    C : set
+        The current set of community nodes.
+    B : set
+        The current set of boundary nodes.
+    T : set of frozenset
+        The current set of boundary edges.
+    I : set of frozenset
+        The current set of internal boundary edges.
+
+    Returns
+    -------
+    R_tmp : float
+        The local modularity after adding node v.
+    B_tmp : set
+        The updated set of boundary nodes.
+    T_tmp : set of frozenset
+        The updated set of boundary edges.
+    I_tmp : set of frozenset
+        The updated set of internal boundary edges.
+    """
+    C_tmp = C | {v}
+    B_tmp = B.copy()
+    T_tmp = T.copy()
+    I_tmp = I.copy()
+    removed_B_nodes = set()
+
+    # Update boundary nodes and edges
+    for neighbour in G[v]:
+        if neighbour not in C_tmp:
+            # v has neighbors not in the community, so it remains a boundary node
+            B_tmp.add(v)
+            # Add edge between v and neighbor to boundary edges
+            T_tmp.add(frozenset([v, neighbour]))
+
+        if neighbour in B:
+            # Check if neighbor should be removed from boundary nodes
+            neighour_still_in_B = False
+            # Go through neighbours neighbours to see if it is still a boundary node
+            for neighbours_neighbour in G[neighbour]:
+                if neighbours_neighbour not in C_tmp:
+                    neighour_still_in_B = True
+                    break
```

Comment:
If I'm reading this right, perhaps it could be simplified with:

```suggestion
            neighour_still_in_B = any(n not in C_tmp for n in G[neighbour])
```

Reply:
Yeah that looks good to me!